### ISSUE TITLE
When triggered, the signal causes an error TypeError: cannot create weak reference to 'unicode' object

### ISSUE THREAD

**Issue by tbazadaykin** [celery/celery#3686]
When triggered, the signal causes an error `TypeError: cannot create weak reference to 'unicode' object.`

```
celery -A myapp report

software -> celery:4.0.1 (latentcall) kombu:4.0.1 py:2.7.9
            billiard:3.5.0.2 py-amqp:2.1.3
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:disabled

broker_url: u'amqp://guest:********@localhost:5672//'
```

**Reproduce**

```
from __future__ import absolute_import, unicode_literals
from celery import Celery

from celery.signals import before_task_publish

app = Celery(
    'myapp',
    broker='amqp://guest@localhost//',
)


@app.task
def add(x, y):
    return x + y

@before_task_publish.connect
def before_task_bublish(sender=None, body=None, **kwargs):
  print 'before_task_publish fired!'

if __name__ == '__main__':
    app.start()

```

In one console start worker:

`python myapp.py worker -l info`

In onther console start `python` and call task:

```
>>> from myapp import add
>>> add.delay(16,16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/venvtest/local/lib/python2.7/site-packages/celery/app/task.py", line 412, in delay
    return self.apply_async(args, kwargs)
  File "/home/venvtest/local/lib/python2.7/site-packages/celery/app/task.py", line 535, in apply_async
    **options
  File "/home/venvtest/local/lib/python2.7/site-packages/celery/app/base.py", line 737, in send_task
    amqp.send_task_message(P, name, message, **options)
  File "/home/venvtest/local/lib/python2.7/site-packages/celery/app/amqp.py", line 547, in send_task_message
    properties=kwargs, retry_policy=retry_policy,
  File "/home/venvtest/local/lib/python2.7/site-packages/celery/utils/dispatch/signal.py", line 222, in send
    self.sender_receivers_cache.get(sender) is NO_RECEIVERS:
  File "/usr/lib/python2.7/weakref.py", line 344, in get
    return self.data.get(ref(key),default)
TypeError: cannot create weak reference to 'unicode' object
```

**Comment by tbazadaykin** [comment]
Works fine if install celery from git:
`pip install git+https://github.com/celery/celery.git`

**Comment by imanhodjaev** [comment]
@ask  @tbazadaykin Also it breaks celery beat with scheduler.

**Comment by ask** [comment]
This is fixed in master, need to release a 4.0.2
